File: construct-binary-tree-from-postorder.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        return build(inorder, 0, inorder.length - 1, postorder, 0,
                postorder.length - 1);
    }

    private TreeNode build(int[] inorder, int l, int r, int[] postorder, int p,
            int q) {
        if (l > r) return null;
        if (l == r) return new TreeNode(postorder[q]);
        TreeNode node = new TreeNode(postorder[q]);
        for (int i = l; i <= r; i++) {
            if (inorder[i] == postorder[q]) {
                node.left = build(inorder, l, i - 1, postorder, p,
                        p + (i - l) - 1);
                node.right = build(inorder, i + 1, r, postorder, p + (i - l),
                        q - 1);
                break;
            }
        }
        return node;
    }
}

File: construct-binary-tree-from-postorder.cc
#include <cstdlib>
#include <vector>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  TreeNode *buildTree(vector<int>& inorder, vector<int>& postorder) {
    return build_tree(inorder, 0, inorder.size() - 1, postorder, 0,
                      postorder.size() - 1);
  }

  TreeNode* build_tree(vector<int>& inorder, int l, int r,
                       vector<int>& postorder, int p, int q) {
    if (l > r || p > q) return NULL;
    for (int i = l; i <= r; i++) {
      if (inorder[i] == postorder[q]) {
        int left_len = i - l;
        TreeNode* root = new TreeNode(postorder[q]);
        root->left = build_tree(inorder, l, i - 1, postorder, p,
                                p + left_len - 1);
        root->right = build_tree(inorder, i + 1, r, postorder, p + left_len,
                                 q - 1);
        return root;
      }
    }
  }
};
